File: services/core-go/internal/agent/task_agent.go
package agent

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"core-go/internal/db"
	"core-go/internal/llm"
)

// --- Agent event types (map 1:1 to sse_payloads.json) ---

// EventKind discriminates the four events the agentic loop can emit.
type EventKind int

const (
	EventText     EventKind = iota // prose token from the LLM
	EventToolCall                  // model requested create_task (UI shows loading)
	EventToolDone                  // task persisted successfully
	EventError                     // validation or DB failure
)

// AgentEvent is one emission from the HandleAgentTask channel.
type AgentEvent struct {
	Kind   EventKind
	Text   string         // EventText: prose token
	Tool   string         // EventToolCall / EventToolDone: tool name
	Args   map[string]any // EventToolCall: validated args (shown in UI)
	TaskID int64          // EventToolDone: Postgres-generated ID
	ErrMsg string         // EventError: human-readable message
}

// --- Schema validation ---

// createTaskArgs mirrors the arguments schema in shared/tools/create_task.json.
// Priority is a string enum — never an integer.
type createTaskArgs struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Priority    string `json:"priority"`
}

// validPriorities is the canonical set from shared/tools/create_task.json.
var validPriorities = map[string]bool{"low": true, "medium": true, "high": true}

var taskIntentHints = []string{
	"create a task",
	"create task",
	"add a task",
	"add task",
	"create a todo",
	"add a todo",
	"new task",
	"todo",
	"to-do",
	"remind me",
	"set a reminder",
	"remember to",
	"add this to my tasks",
	"task:",
}

var taskQueryHints = []string{
	"my task",
	"my tasks",
	"list tasks",
	"show tasks",
	"show my tasks",
	"what are my tasks",
	"any task",
	"pending task",
	"pending tasks",
	"today task",
	"today tasks",
	"do i have any tasks",
}

var taskIntentActionWords = []string{
	"create",
	"add",
	"set",
	"remember",
	"remind",
	"track",
	"schedule",
	"note",
	"record",
}

var taskIntentSubjectWords = []string{
	"task",
	"todo",
	"to-do",
	"reminder",
	"deadline",
	"follow-up",
	"follow up",
}

var nonTaskQuestionPrefixes = []string{
	"what",
	"why",
	"how",
	"who",
	"where",
	"when",
	"explain",
	"tell me",
	"can you explain",
}

func looksLikeTaskIntent(userMessage string) bool {
	lc := strings.ToLower(strings.TrimSpace(userMessage))
	if lc == "" {
		return false
	}

	for _, hint := range taskIntentHints {
		if strings.Contains(lc, hint) {
			return true
		}
	}

	if strings.HasPrefix(lc, "task ") || strings.HasPrefix(lc, "todo ") {
		return true
	}

	startsLikeQuestion := false
	for _, prefix := range nonTaskQuestionPrefixes {
		if strings.HasPrefix(lc, prefix+" ") || lc == prefix {
			startsLikeQuestion = true
			break
		}
	}

	hasActionWord := false
	for _, word := range taskIntentActionWords {
		if strings.Contains(lc, word) {
			hasActionWord = true
			break
		}
	}

	hasSubjectWord := false
	for _, word := range taskIntentSubjectWords {
		if strings.Contains(lc, word) {
			hasSubjectWord = true
			break
		}
	}

	if hasActionWord && hasSubjectWord {
		return true
	}

	if startsLikeQuestion {
		return false
	}

	return false
}

func looksLikeTaskQuery(userMessage string) bool {
	lc := strings.ToLower(strings.TrimSpace(userMessage))
	if lc == "" {
		return false
	}

	for _, hint := range taskQueryHints {
		if strings.Contains(lc, hint) {
			return true
		}
	}

	if strings.HasPrefix(lc, "task ") || strings.HasPrefix(lc, "tasks ") {
		return true
	}

	return false
}

// ShouldUseTaskAgent decides if a user query should be routed to the task
// agent pipeline (task creation or task list/query), optionally forced by UI.
func ShouldUseTaskAgent(userMessage string, forceTask bool) bool {
	if forceTask {
		return true
	}
	return looksLikeTaskIntent(userMessage) || looksLikeTaskQuery(userMessage)
}

func validateCreateTaskArgs(raw json.RawMessage) (createTaskArgs, error) {
	var args createTaskArgs
	if err := json.Unmarshal(raw, &args); err != nil {
		return args, fmt.Errorf("unmarshal args: %w", err)
	}
	if strings.TrimSpace(args.Title) == "" {
		return args, fmt.Errorf("'title' is required and must be non-empty")
	}
	if args.Priority == "" {
		args.Priority = "medium" // schema default
	}
	if !validPriorities[args.Priority] {
		return args, fmt.Errorf("'priority' must be one of low|medium|high, got %q", args.Priority)
	}
	return args, nil
}

// --- System prompt ---

const agentSystemPrompt = `You are a personal task management assistant.
When the user wants to create, add, or record a task, use the create_task tool.
Extract the task title (required), description (if mentioned), and priority
(if mentioned; must be "low", "medium", or "high"; default "medium").
If the user's intent is not to create a task, respond conversationally without using a tool.`

// --- TaskAgent ---

// TaskAgent runs the agentic loop that detects task-creation intent,
// executes the tool, and generates a final summary for the user.
type TaskAgent struct {
	repo db.TaskRepository
}

// NewTaskAgent returns a TaskAgent backed by the given repository.
func NewTaskAgent(repo db.TaskRepository) *TaskAgent {
	return &TaskAgent{repo: repo}
}

// HandleAgentTask runs the full agentic loop for userMessage and returns a
// read-only channel of AgentEvents. The channel is closed when the loop
// completes or ctx is cancelled.
//
// userID is the device-generated UUID of the requesting user. It is stored
// alongside the task so tasks are per-user. Pass "admin" for system tasks.
//
//  1. Checks whether userMessage is explicit task intent.
//  2. If yes, sends userMessage to Ollama with the create_task tool attached.
//     If not, sends userMessage without tools for normal conversational chat.
//  2. If Ollama returns a ToolCall chunk:
//     a. Validates the extracted args (title required, priority enum).
//     b. Emits EventToolCall so the UI can show a loading state.
//     c. Calls TaskRepository.CreateTask with userID.
//     d. Emits EventToolDone with the generated task ID.
//     e. Sends a tool-result confirmation back to Ollama for a final summary.
//  3. Streams all LLM text tokens as EventText.
func (ta *TaskAgent) HandleAgentTask(ctx context.Context, userMessage, userID string, forceTask bool) (<-chan AgentEvent, error) {
	if looksLikeTaskQuery(userMessage) && !forceTask {
		return ta.handleTaskListQuery(ctx, userID)
	}

	messages := []llm.Message{
		{Role: "system", Content: agentSystemPrompt},
		{Role: "user", Content: userMessage},
	}

	var tools []llm.Tool
	if forceTask || looksLikeTaskIntent(userMessage) {
		tools = []llm.Tool{llm.CreateTaskTool}
	}

	ch, err := llm.StreamChat(ctx, messages, tools)
	if err != nil {
		return nil, fmt.Errorf("agent: start stream: %w", err)
	}

	out := make(chan AgentEvent, 16)
	go ta.runLoop(ctx, ch, messages, userID, out)
	return out, nil
}

func (ta *TaskAgent) handleTaskListQuery(ctx context.Context, userID string) (<-chan AgentEvent, error) {
	tasks, err := ta.repo.ListTasks(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("agent: list tasks: %w", err)
	}

	message := "You don't have any tasks yet."
	if len(tasks) > 0 {
		limit := len(tasks)
		if limit > 5 {
			limit = 5
		}

		lines := make([]string, 0, limit+2)
		lines = append(lines, "Here are your tasks:")
		for i := 0; i < limit; i++ {
			t := tasks[i]
			lines = append(lines, fmt.Sprintf("%d) %s [%s | %s]", i+1, t.Title, t.Status, t.Priority))
		}
		if len(tasks) > limit {
			lines = append(lines, fmt.Sprintf("...and %d more.", len(tasks)-limit))
		}
		message = strings.Join(lines, "\n")
	}

	out := make(chan AgentEvent, 1)
	go func() {
		defer close(out)
		emit(ctx, out, AgentEvent{Kind: EventText, Text: message})
	}()

	return out, nil
}

// runLoop reads from the first-turn Chunk channel and orchestrates the
// validation → DB write → second-turn summary flow.
func (ta *TaskAgent) runLoop(
	ctx context.Context,
	ch <-chan llm.Chunk,
	firstTurnMessages []llm.Message,
	userID string,
	out chan<- AgentEvent,
) {
	defer close(out)

	for chunk := range ch {
		switch chunk.Kind {

		case llm.KindText:
			emit(ctx, out, AgentEvent{Kind: EventText, Text: chunk.Text})

		case llm.KindToolCall:
			tc := chunk.ToolCall

			// Step 2a — validate args against the create_task schema.
			args, err := validateCreateTaskArgs(tc.Arguments)
			if err != nil {
				emit(ctx, out, AgentEvent{
					Kind:   EventError,
					ErrMsg: fmt.Sprintf("tool arg validation: %v", err),
				})
				return
			}

			validatedArgs := map[string]any{
				"title":       args.Title,
				"description": args.Description,
				"priority":    args.Priority,
			}

			// Step 2b — emit tool_call so the UI shows a loading state.
			emit(ctx, out, AgentEvent{
				Kind: EventToolCall,
				Tool: tc.Name,
				Args: validatedArgs,
			})

			// Step 2c — execute TaskRepository.CreateTask, scoped to the requesting user.
			taskID, err := ta.repo.CreateTask(ctx, args.Title, args.Description, args.Priority, userID)
			if err != nil {
				emit(ctx, out, AgentEvent{
					Kind:   EventError,
					ErrMsg: fmt.Sprintf("create task: %v", err),
				})
				return
			}

			// Step 2d — emit tool_done with the Postgres-generated ID.
			emit(ctx, out, AgentEvent{
				Kind:   EventToolDone,
				Tool:   tc.Name,
				TaskID: int64(taskID),
			})

			// Step 2e — build second-turn history and stream the final summary.
			ta.streamSummary(ctx, firstTurnMessages, tc.Name, validatedArgs, int64(taskID), out)
			return // agentic loop ends after one tool execution
		}
	}
}

// streamSummary reconstructs the full message history including the tool
// result and streams Ollama's final natural-language confirmation.
func (ta *TaskAgent) streamSummary(
	ctx context.Context,
	firstTurnMessages []llm.Message,
	toolName string,
	validatedArgs map[string]any,
	taskID int64,
	out chan<- AgentEvent,
) {
	fallbackText := fmt.Sprintf("Task created successfully (ID: %d).", taskID)

	// Reconstruct the assistant's tool-call message for Ollama's history.
	toolCallsJSON, _ := json.Marshal([]map[string]any{{
		"function": map[string]any{
			"name":      toolName,
			"arguments": validatedArgs,
		},
	}})

	// Tool result sent back to the model as the "tool" role message.
	toolResult, _ := json.Marshal(map[string]any{
		"status":  "success",
		"task_id": taskID,
		"title":   validatedArgs["title"],
	})

	// Build a fresh slice to avoid mutating the original firstTurnMessages.
	followUp := append(
		append([]llm.Message{}, firstTurnMessages...),
		llm.Message{Role: "assistant", Content: "", ToolCalls: toolCallsJSON},
		llm.Message{Role: "tool", Content: string(toolResult)},
	)

	summaryCh, err := llm.StreamChat(ctx, followUp, nil)
	if err != nil {
		emit(ctx, out, AgentEvent{Kind: EventText, Text: fallbackText})
		return
	}

	emittedText := false
	for sc := range summaryCh {
		if sc.Kind == llm.KindText {
			emittedText = true
			emit(ctx, out, AgentEvent{Kind: EventText, Text: sc.Text})
		}
	}

	if !emittedText {
		emit(ctx, out, AgentEvent{Kind: EventText, Text: fallbackText})
	}
}

// emit sends e to ch while respecting ctx cancellation.
func emit(ctx context.Context, ch chan<- AgentEvent, e AgentEvent) {
	select {
	case ch <- e:
	case <-ctx.Done():
	}
}

File: services/core-python/app/agent/task_agent.py
import json
from dataclasses import dataclass, field
from typing import Any, AsyncGenerator, Literal

from app.agent.llm import CREATE_TASK_TOOL, stream_chat
from app.database.task_repository import TaskRepository

# --- Agent event types (map 1:1 to sse_payloads.json) ---


@dataclass
class TextEvent:
    """Prose token emitted by the LLM."""
    kind: Literal["text"] = field(default="text", init=False)
    content: str = ""


@dataclass
class ToolCallEvent:
    """Model requested create_task — UI should show a loading state."""
    kind: Literal["tool_call"] = field(default="tool_call", init=False)
    tool: str = ""
    args: dict[str, Any] = field(default_factory=dict)


@dataclass
class ToolDoneEvent:
    """Task successfully persisted in Postgres."""
    kind: Literal["tool_done"] = field(default="tool_done", init=False)
    tool: str = ""
    task_id: int = 0


@dataclass
class ErrorEvent:
    """Validation or DB failure — loop terminates after this."""
    kind: Literal["error"] = field(default="error", init=False)
    message: str = ""


AgentEvent = TextEvent | ToolCallEvent | ToolDoneEvent | ErrorEvent

# --- System prompt ---

_AGENT_SYSTEM_PROMPT = (
    "You are a personal task management assistant.\n"
    "When the user wants to create, add, or record a task, use the create_task tool.\n"
    "Extract the task title (required), description (if mentioned), and priority "
    "(if mentioned; 0=low 1=medium 2=high 3=urgent; default 0).\n"
    "If the user's intent is not to create a task, respond conversationally "
    "without using a tool."
)

# --- Schema validation ---


def _validate_args(args: dict[str, Any]) -> tuple[str, str, int]:
    """Validate tool arguments against the create_task schema.

    Returns:
        (title, description, priority) on success.

    Raises:
        ValueError with a descriptive message on any violation.
    """
    title = args.get("title", "")
    if not isinstance(title, str) or not title.strip():
        raise ValueError("'title' is required and must be a non-empty string")

    description = args.get("description", "")
    if not isinstance(description, str):
        raise ValueError("'description' must be a string")

    priority = args.get("priority", 0)
    if not isinstance(priority, int) or not (0 <= priority <= 3):
        raise ValueError(f"'priority' must be an integer 0–3, got {priority!r}")

    return title.strip(), description, priority


# --- Agentic loop ---


async def handle_agent_task(
    user_message: str,
    task_repo: TaskRepository,
) -> AsyncGenerator[AgentEvent, None]:
    """Full agentic loop for task-creation intent.

    Steps:
        1. Sends user_message to Ollama with the create_task tool attached.
        2. If Ollama returns a tool_call chunk:
           a. Validates the extracted args (title required, priority 0–3).
           b. Yields ToolCallEvent so the UI can show a loading state.
           c. Calls TaskRepository.create_task.
           d. Yields ToolDoneEvent with the Postgres-generated task ID.
           e. Sends a tool-result confirmation back to Ollama and streams
              the model's final natural-language summary.
        3. Streams all LLM prose tokens as TextEvent.

    Yields:
        TextEvent, ToolCallEvent, ToolDoneEvent, or ErrorEvent.
    """
    messages: list[dict[str, Any]] = [
        {"role": "system", "content": _AGENT_SYSTEM_PROMPT},
        {"role": "user", "content": user_message},
    ]

    async for chunk in stream_chat(messages, tools=[CREATE_TASK_TOOL]):

        if chunk.kind == "text":
            yield TextEvent(content=chunk.content)

        elif chunk.kind == "tool_call":

            # Step 2a — validate args against the create_task schema.
            try:
                title, description, priority = _validate_args(chunk.args)
            except ValueError as exc:
                yield ErrorEvent(message=f"tool arg validation: {exc}")
                return

            validated_args = {
                "title": title,
                "description": description,
                "priority": priority,
            }

            # Step 2b — emit tool_call so the UI shows a loading state.
            yield ToolCallEvent(tool=chunk.name, args=validated_args)

            # Step 2c — execute TaskRepository.create_task.
            try:
                task_id = await task_repo.create_task(title, description, priority)
            except Exception as exc:
                yield ErrorEvent(message=f"create task: {exc}")
                return

            # Step 2d — emit tool_done with the Postgres-generated ID.
            yield ToolDoneEvent(tool=chunk.name, task_id=task_id)

            # Step 2e — second turn: feed tool result back to Ollama and
            # stream the model's final natural-language confirmation.
            tool_result = json.dumps({
                "status": "success",
                "task_id": task_id,
                "title": title,
            })

            followup: list[dict[str, Any]] = [
                *messages,
                {
                    "role": "assistant",
                    "content": "",
                    "tool_calls": [
                        {"function": {"name": chunk.name, "arguments": validated_args}}
                    ],
                },
                {"role": "tool", "content": tool_result},
            ]

            async for summary in stream_chat(followup):
                if summary.kind == "text":
                    yield TextEvent(content=summary.content)

            return  # loop ends after one tool execution
